### PR TITLE
Posts: Avoid the use of globals in get_the_content() and related functions #1082 

### LINKED ISSUE #1082: Posts: Avoid the use of globals in get_the_content() and related functions
I'm opening this issue so it serves as a base for a future PR that I will submit after #1051 gets merged, but since I already tested this and it fixes the problem, I'm leaving it here so it doesn't get lost.

This is a fix for [WP 42814](https://core.trac.wordpress.org/ticket/42814).

The changesets that fix this are [#44941](https://core.trac.wordpress.org/changeset/44941) and [#48114](https://core.trac.wordpress.org/changeset/48114).

### PR DESCRIPTION
**Pull request by mattyrob**
## Description
Fixes #1082 and a PHP 7.2 Warning

Backport of changesets that fix this are [#44941](https://core.trac.wordpress.org/changeset/44941) and [#48114](https://core.trac.wordpress.org/changeset/48114).

This is a rework of PR #1083

## Motivation and context
Backport for PHP compatibility.

## How has this been tested?
Comes with unit tests.

## Screenshots
N/A

## Types of changes
- Bug fix

### CODE REVIEW

**Comment by bahiirwa** [comment]
Looking at this PR. Is https://core.trac.wordpress.org/changeset/51462/ too early for this?

File: src/wp-includes/class-wp-query.php [hunk 1]
```diff
@@ -4164,6 +4164,53 @@ public function setup_postdata( $post ) {
...
+		 */
+		do_action_ref_array( 'the_post', array( &$post, &$this ) );
+
+		return true;
+	}
+
+	/**
+	 * Generate post data.
+	 *
+	 * @since 5.2.0
```
Comment (xxsimoxx) [inline src/wp-includes/class-wp-query.php:4199]: Missing `WP-`.

File: src/wp-includes/query.php [hunk 2]
```diff
@@ -1009,3 +1009,23 @@ function setup_postdata( $post ) {
 
 	return false;
 }
+
+/**
+ * Generates post data.
+ *
+ * @since 5.2.0
```
Comment (xxsimoxx) [inline src/wp-includes/query.php:1016]: Missing `WP-`.

File: src/wp-includes/class-wp-query.php (hunk 1 above)
Comment (mattyrob) [reply src/wp-includes/class-wp-query.php]: PR updated.

**Review by bahiirwa — APPROVED** [review]
Clean change. LGTM

**Review by xxsimoxx — APPROVED** [review]
Tested for some days on a live site.
Also tested `get_the_content` out of the loop.
Looks good.

**Comment by mattyrob** [comment]
Has 2 reviews, merging.

**Pull request merged**